### ISSUE TITLE
False positive union-attr error when iterating over union of two iterables

### ISSUE THREAD

**Issue by brownan** [python/mypy#14103]
**Bug Report**

When iterating over the union of two iterable classes, where one class implements an `__iter__` and `__next__`, but the other uses a generator function as its `__iter__`, mypy produces the error 
```
iter_test.py:31: error: Item "Class2" of "Union[Class2, Iterator[Tuple[Any, ...]]]" has no attribute "__next__"  [union-attr]
```

**To Reproduce**
Here's an example: https://mypy-play.net/?mypy=latest&python=3.10&gist=a5f374765c77567f926505e07717282d

```python
from typing import Any, Iterator, Tuple, TypeVar, Union

Self = TypeVar("Self")


class Class1:
    def __iter__(self: Self) -> Self:
        return self

    def __next__(self) -> Tuple:
        raise StopIteration


class Class2:
    def __iter__(self) -> Iterator[Tuple[Any, ...]]:
        yield (1,)


def test1(m: Class1) -> None:
    for i in m: # Works
        print(i)


def test2(m: Class2) -> None:
    for i in m: # Works
        print(i)


def test3(m: Union[Class1, Class2]) -> None:
    # error: Item "Class2" of "Union[Class2, Iterator[Tuple[Any, ...]]]" has no attribute "__next__"  [union-attr]
    for i in m:
        print(i)
```

**Expected Behavior**

Expected mypy to recognize both items in the union to support iteration.

**Actual Behavior**

```
iter_test.py:31: error: Item "Class2" of "Union[Class2, Iterator[Tuple[Any, ...]]]" has no attribute "__next__"  [union-attr]
```

**Your Environment**

- Mypy version used: tested against both 0.991 and 0.982
- Mypy command-line flags: none
- Mypy configuration options from `mypy.ini` (and other config files): no config
- Python version used: 3.8.10

**Comment by ShadowLNC** [comment]
I have what I think is the same issue, I can reproduce using this example:

```python3
a = {"a": "b", "c": 2}  # In our application, this dictionary is dynamic and may be empty.
for k, v in a.items() or zip(["a", "x", "y"], repeat("")):
    print(k, v)
```

(Our actual code iterates over `myDict.items() or zip(set_or_tuple_of_str, repeat(...))`, where the dictionary values are of a custom type.)


Running Python 3.11, `mypy` version `0.991` gives the error:
```
Item "dict_items[str, object]" of "Union[Iterator[Tuple[str, object]], dict_items[str, object]]" has no attribute "__next__"  [union-attr]
```

This appears to be a regression. Our code previously passed static analysis using Python 3.8 and `mypy` version `0.910`.

**Comment by ilevkivskyi** [comment]
This works correctly on the current master.
